File: CMPS12B: Introduction to Data Structures/pa5: Dictionary Client/Dictionary.c
/** ***************************************************************************
 * Name: Juan D. Cardozo
 * Student I.D: 1362585
 * CRUZ I.D: jcardozo
 * Class: CMPS 12B
 * Section: Mondays 6-8
 * Date: 05/09/2015
 * Dictionary.c
 * It implements Dictionary.h
 ******************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "Dictionary.h"

//Hash Functions and variables-------------------------------------------------

// Size of Hash table
const int tableSize = 101;

// Hash Functions
// rotate_left()
// rotate the bits in an unsigned int
unsigned int rotate_left(unsigned int value, int shift) {
	int sizeInBits = 8*sizeof(unsigned int);
	shift = shift & (sizeInBits - 1);
	if ( shift == 0 )
		return value;
	      return (value << shift) | (value >> (sizeInBits - shift));
}

// pre_hash()
// turn a string into an unsigned int
unsigned int pre_hash(char* input) {
	unsigned int result = 0xBAE86554;
	while (*input) {
		result ^= *input++;
		result = rotate_left(result, 5);
	}
	return result;
}

// hash()
// turns a string into an int in the range 0 to tableSize-1
int hash(char* key){
	return pre_hash(key)%tableSize;
}

// private types --------------------------------------------------------------

// NodeObj
typedef struct NodeObj {
    char* key;
    char* value;
    struct NodeObj* next;
} NodeObj;

// Node
typedef NodeObj* Node;

// newNode()
// constructor of the Node type
Node newNode(char* x, char* y)
{
    Node N = malloc(sizeof (NodeObj));
    assert(N != NULL);
    N->key = x;
    N->value = y;
    N->next = NULL;
    return (N);
}

// freeNode()
// destructor for the Node type

void freeNode(Node* pN)
{
    if (pN != NULL && *pN != NULL) {
        free(*pN);
        *pN = NULL;
    }
}

// Dictionary
typedef struct DictionaryObj {
    Node* table;
    int numItems;
}DictionaryObj;


// newDictionary()
// constructor for the Dictionary type
Dictionary newDictionary(void)
{
  Dictionary D = malloc(sizeof(DictionaryObj));
  D->table = calloc(tableSize,sizeof(Node));
  D-> numItems = 0;
	return D;
}

// freeDictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* pD)
{
	 if (pD != NULL && *pD != NULL) {
        if (!isEmpty(*pD)) makeEmpty(*pD);
	free((*pD)->table);
        free(*pD);
        *pD = NULL;
    }
}
	

// isEmpty()
// returns 1 (true) if S is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D)
{
	return (D->numItems == 0)? 1:0;
}

// size()
// returns the number of (key, value) pairs in D
// pre: none
int size(Dictionary D)
{
	return (D->numItems);
}

// lookup()
// returns the value v such that (k, v) is in D, or returns NULL if no 
// such value v exists.
// pre: none
char* lookup(Dictionary D, char* k)
{
	int  index = hash(k);
	Node N = D->table[index];
	if(N == NULL){
		return NULL;
	}else if(N->next == NULL){
		return N->value;
	}else{ // N->next != NULL
		while(strcmp(k,N->key) != 0){
			N = N->next;
		}
		return N->value;
	}
}

// insert()
// inserts new (key,value) pair into D
// pre: lookup(D, k)==NULL
void insert(Dictionary D, char* k, char* v)
{
  if (lookup(D,k) != NULL) {
        fprintf(stderr, "Dictionary Error: cannot insert dublicate of keys\n");
        exit(EXIT_FAILURE);
    }
	int index = hash(k);
	Node N = D->table[index];
	if(N == NULL){
	        D->table[index] = newNode(k,v);
	}else{ // N->next != NULL
		while(N->next != NULL){
			N = N->next;
		}
		N = newNode(k,v);
		N->next = D->table[index];
		D->table[index] = N;
	}
	D->numItems++;
}

// delete()
// deletes pair with the key k
// pre: lookup(D, k)!=NULL
void delete(Dictionary D, char* k)
{
	if (lookup(D, k) ==NULL) {
        fprintf(stderr, "Dictionary Error: calling delete() on Non-Existent key\n");
        exit(EXIT_FAILURE);
    }
	int index = hash(k);
	Node N = D->table[index];
	if(N->next == NULL){
	  freeNode(&N);
	  D->table[index] = NULL;
	}else{ // N->next != NULL
		while(strcmp(k,N->key) != 0){
			N = N->next;
		}
		Node G = N->next;
		N->next = N->next->next;
		freeNode(&G);
	}
	D->numItems--;
}

// makeEmpty()
// re-sets D to the empty state.
// pre: none
void makeEmpty(Dictionary D)
{
  Node N;
  Node G;
   for(int i = 0; i < tableSize; i++){
     N = D->table[i];
     while(N != NULL){
       G = N;
       N = N->next;
       freeNode(&G);
     }
     if(D->table[i] != NULL){
       D->table[i] = NULL;
     }
   }
   D->numItems = 0;
}

// printDictionary()
// pre: none
// prints a text representation of D to the file pointed to by out
void printDictionary(FILE* out, Dictionary D)
{
  Node N;
	for(int i = 0; i < tableSize; i++){
	  if(D->table[i] != NULL){
	    N = D->table[i];
	    while(N != NULL){
		fprintf(out, "%s %s\n", N->key, N->value);
		N = N->next;
	    }
	  }	
	
	}

}

File: CMPS12B: Introduction to Data Structures/pa3: Dictionary Search/Dictionary.java
/** ***************************************************************************
* Name: Juan D. Cardozo
* Student I.D: 1362585
* CRUZ I.D: jcardozo
* Class: CMPS 12M
* Section: Mondays 6-8
* Date: 04/13/2015
* Dictionary.java
* Creates a Dictionary with functions to alter it
*****************************************************************************/

public class Dictionary implements DictionaryInterface{

    // Private fields for dictionary
    private Node head;
    private int  numItems;

    //Dictionary Constructor
    Dictionary(){
	head = null;
	numItems =0;
    }

    // Node Class Constructor  
    private class Node{
	
		Node next;
		
		String key = "";
		String value = "";
                
	        // Node Constructor
		Node(String x,String y){
		   key +=x;
		   value +=y;
		   next = null;
		}
    }

    // Node findKey();
    // Return a reference to the Node containing its argument key,
    private Node findKey(String key){
		Node N = head;
		for(int i = numItems; i > 0; i-- ){
		    if(key.compareTo(N.key) == 0){
					return N;
				}
				N = N.next;
			}
		return null;
    }

    // isEmpty()
    // Returns true if there are no pairs, else false;
    public boolean isEmpty(){
		return (numItems == 0);
    }

    // size()
    // Returns the size of the list
    public int size(){
		return numItems;
    }

    // lookup()
    // Return value field if key matches the argument key, otherwise null
    public String lookup(String key){
		Node N = findKey(key);
		if(N != null){
			return N.value;
		}else{ // if key == null
			return null;
		}
    }	

    // insert()
    // inserts a node at the begining of the list
    public void insert(String key, String value){
	if(findKey(key) != null){
	    throw new  KeyCollisionException("cannot insert duplicate keys");
	}else{
	    Node newNode = new Node(key,value);
	    newNode.next = head;
	    head = newNode;
	    numItems++;
	}
    }

    // delete()
    // deletes the node with the spcified key
    public void delete(String key){
	if(findKey(key) ==  null){
	    throw new KeyNotFoundException("cannot delete non-existent key");
	}else{
	    Node N = head;
	    Node temp = N;
	    Node P  = findKey(key);
	    while(N.next != P){
		  N = N.next;
	    }
	    P = N;
  	    Node C = P.next;
	    P.next = C.next;
	    C.next = null;
	    numItems--;
	}
    }

    // makeEmpty()
    // Deletes everything in the list
    public void makeEmpty(){
		head = null;
		numItems = 0;
    }

    // toString()
    // Make into printable list
    public String toString(){
		String s ="";
		Node N = head;
		int listLength = size();
		while(listLength--  > 0){
		    
		    s=N.key+" "+N.value+"\n"+s;
		    N = N.next;
		}
		return s;
    }
}